pkg/tun/ip/v4header.go

```diff
@@ -0,0 +1,269 @@
+
+func (h V4Header) Version() int {
+	return int(h[0] >> 4)
+}
+
+func (h V4Header) HeaderLen() int {
+	return int(h[0]&0x0f) * 4
+}
+
+func (h V4Header) SetHeaderLen(hl int) {
+	h[0] = (h[0] & 0xf0) | uint8(hl/4)
+}
+
+func (h V4Header) DSCP() int {
+	return int(h[1] >> 2)
+}
+
+func (h V4Header) ECN() int {
+	return int(h[1] & 0x3)
+}
+
+func (h V4Header) PayloadLen() int {
+	return int(binary.BigEndian.Uint16(h[2:]) - uint16(h.HeaderLen()))
+}
+
+func (h V4Header) SetPayloadLen(len int) {
+	binary.BigEndian.PutUint16(h[2:], uint16(len+h.HeaderLen()))
+}
+
+func (h V4Header) ID() uint16 {
+	return binary.BigEndian.Uint16(h[4:])
+}
+
+func (h V4Header) SetID(id int) {
+	binary.BigEndian.PutUint16(h[4:], uint16(id))
+}
+
+func (h V4Header) Flags() ipv4.HeaderFlags {
+	return ipv4.HeaderFlags(h[6]) >> 5
+}
+
+func (h V4Header) SetFlags(flags ipv4.HeaderFlags) {
+	h[6] = (h[6] & 0x1f) | uint8(flags<<5)
+}
+
+func (h V4Header) FragmentOffset() int {
+	return int(binary.BigEndian.Uint16(h[6:]) & 0x1fff)
+}
+
+func (h V4Header) SetFragmentOffset(fragOff int) {
+	flagBits := uint16(h[6]&0b11100000) << 8
+	binary.BigEndian.PutUint16(h[6:], flagBits|uint16(fragOff&0x1fff))
+}
+
+func (h V4Header) TTL() int {
+	return int(h[8])
+}
+
+func (h V4Header) SetTTL(hops int) {
+	h[8] = uint8(hops)
+}
+
+func (h V4Header) L4Protocol() int {
+	return int(h[9])
+}
+
+func (h V4Header) SetL4Protocol(proto int) {
+	h[9] = uint8(proto)
+}
+
+func (h V4Header) Checksum() int {
+	return int(binary.BigEndian.Uint16(h[10:]))
+}
+
+func (h V4Header) Source() net.IP {
+	return net.IP(h[12:16])
+}
+
+func (h V4Header) Destination() net.IP {
+	return net.IP(h[16:20])
+}
+
+func (h V4Header) SetSource(ip net.IP) {
+	if ip4 := ip.To4(); ip4 != nil {
+		copy(h[12:], ip)
+	}
+}
+
+func (h V4Header) SetDestination(ip net.IP) {
+	if ip4 := ip.To4(); ip4 != nil {
+		copy(h[16:20], ip)
+	}
+}
+
+func (h V4Header) Options() ([]V4Option, error) {
+	optBytes := h[20:h.HeaderLen()]
+	var opts []V4Option
+	obl := len(optBytes)
+	for i := 0; i < obl; {
+		optionType := optBytes[i]
+		switch optionType {
+		case 0: // end of list
+			i = obl
+		case 1: // byte padding
+			opts = append(opts, V4Option(optBytes[i:i+1]))
+			i++
+		default:
+			if i+1 < obl {
+				optLen := int(optBytes[i+1])
+				if i+optLen < obl {
+					opts = append(opts, V4Option(optBytes[i:i+optLen]))
+					i += optLen
+					continue
+				}
+			}
+			return nil, errors.New("option data is outside IPv4 header")
+		}
+	}
+	return opts, nil
+}
+
+func (h V4Header) Packet() []byte {
+	return h[:h.HeaderLen()+h.PayloadLen()]
+}
+
+func (h V4Header) Payload() []byte {
+	return h[h.HeaderLen() : h.HeaderLen()+h.PayloadLen()]
+}
+
+func (h V4Header) SetChecksum() {
+	h[10] = 0 // clear current checksum
+	h[11] = 0
+
+	s := 0
+	t := h.HeaderLen()
+	for i := 0; i < t; i += 2 {
+		s += int(h[i])<<8 | int(h[i+1])
+	}
+	for s > 0xffff {
+		s = (s >> 16) + (s & 0xffff)
+	}
+	c := ^uint16(s)
+	binary.BigEndian.PutUint16(h[10:], c)
+}
```

Comment:
I think this will panic on an odd-length header?

Reply:
It would, but there will never be an odd-length IPv4-header. Without extension headers, the IPv4 header is 20 bytes (or 5 * groups of 32-bit). It can be extended by more groups of 4, but never with odd bytes.

Reply:
You're right. I was thrown off by RFC 1071 specifically handling odd-length inputs, but that clearly doesn't apply to IPv4 headers.